File: yurt-jdbc/src/main/java/com/github/liulus/yurt/jdbc/SQLExecutorImpl.java
package com.github.liulus.yurt.jdbc;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.BeanUtils;
import org.springframework.dao.IncorrectResultSizeDataAccessException;
import org.springframework.jdbc.core.ConnectionCallback;
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcOperations;
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
import org.springframework.jdbc.core.namedparam.SqlParameterSource;
import org.springframework.jdbc.support.GeneratedKeyHolder;
import org.springframework.jdbc.support.KeyHolder;
import org.springframework.util.Assert;
import org.springframework.util.CollectionUtils;
import org.springframework.util.ReflectionUtils;
import org.springframework.util.StringUtils;

import javax.sql.DataSource;
import java.beans.PropertyDescriptor;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

import static com.github.liulus.yurt.jdbc.SQLBuilder.sqlParams;

/**
 * @author liulu
 * @version V1.0
 * @since 2020/11/10
 */
public class SQLExecutorImpl implements SQLExecutor {

    private static final Logger logger = LoggerFactory.getLogger(SQLExecutorImpl.class);

    private String dbName;

    private final NamedParameterJdbcOperations namedParameterJdbcOperations;

    public SQLExecutorImpl(DataSource dataSource) {
        this.namedParameterJdbcOperations = new NamedParameterJdbcTemplate(dataSource);
    }

    public SQLExecutorImpl(NamedParameterJdbcOperations namedParameterJdbcOperations) {
        this.namedParameterJdbcOperations = namedParameterJdbcOperations;
    }


    @Override
    public <E> long insert(E entity) {
        Assert.notNull(entity, "insert with entity can not be null");
        Class<?> entityClass = entity.getClass();
        String sql = SQLBuilder.insertSQL(entity).toString();
        TableMetadata tableMetadata = TableMetadata.forClass(entityClass);
        KeyHolder keyHolder = new GeneratedKeyHolder();
        SqlParameterSource sqlParams = sqlParams(entity);
        logger.debug("insert sql: {} \n params: {}", sql, sqlParams);
        namedParameterJdbcOperations.update(sql, sqlParams, keyHolder);
        long id = Objects.requireNonNull(keyHolder.getKey()).longValue();
        PropertyDescriptor keyPs = BeanUtils.getPropertyDescriptor(entityClass, tableMetadata.getIdField());
        // set key property
        if (keyPs != null) {
            ReflectionUtils.invokeMethod(keyPs.getWriteMethod(), entity, id);
        }
        return id;
    }

    @Override
    public <E> int batchInsert(Collection<E> eList) {
        if (CollectionUtils.isEmpty(eList)) {
            return 0;
        }
        E entity = eList.iterator().next();
        SQL sql = SQLBuilder.insertSQL(entity);

        SqlParameterSource[] parameterSources = eList.stream()
                .map(SQLBuilder::sqlParams).toArray(SqlParameterSource[]::new);
        logger.debug("batchInsert sql: {} \n params: {}", sql, parameterSources);
        int[] updateResult = namedParameterJdbcOperations.batchUpdate(sql.toString(), parameterSources);

        int row = 0;
        for (int res : updateResult) {
            row += res;
        }
        return row;
    }

    @Override
    public <E> int updateIgnoreNull(E entity) {
        Assert.notNull(entity, "update with entity can not be null");
        String sql = SQLBuilder.updateSQL(entity, true).toString();
        SqlParameterSource sqlParams = sqlParams(entity);
        logger.debug("update sql: {} \n params: {}", sql, sqlParams);
        return namedParameterJdbcOperations.update(sql, sqlParams);
    }

    @Override
    public <E> int deleteById(Class<E> eClass, Long id) {
        Assert.notNull(id, "delete id can not be null");
        String sql = SQLBuilder.deleteSQL(eClass).toString();
        SqlParameterSource sqlParams = sqlParams(id);
        logger.debug("delete sql: {} \n params: {}", sql, sqlParams);
        return namedParameterJdbcOperations.update(sql, sqlParams);
    }

    @Override
    public <E> int deleteLogicalById(Class<E> eClass, Long id) {
        Assert.notNull(id, "delete logical id can not be null");
        String sql = SQLBuilder.deleteLogicalSQL(eClass, getDbName()).toString();
        SqlParameterSource sqlParams = sqlParams(id);
        logger.debug("delete sql: {} \n params: {}", sql, sqlParams);
        return namedParameterJdbcOperations.update(sql, sqlParams);
    }

    @Override
    public <E> E selectById(Class<E> eClass, Long id) {
        Assert.notNull(id, "id can not be null");
        SQL sql = SQLBuilder.selectByIdSQL(eClass);
        return selectForObject(sql, id, eClass);
    }

    @Override
    public <E> List<E> selectByIds(Class<E> eClass, Collection<Long> ids) {
        if (CollectionUtils.isEmpty(ids)) {
            return Collections.emptyList();
        }
        SQL sql = SQLBuilder.selectByIdsSQL(eClass);
        return selectForList(sql, ids, eClass);
    }

    @Override
    public <E> E selectForObject(SQL sql, Object params, Class<E> requiredType) {
        List<E> rs = selectForList(sql, params, requiredType);
        if (CollectionUtils.isEmpty(rs)) {
            return null;
        }
        if (rs.size() > 1) {
            throw new IncorrectResultSizeDataAccessException(1, rs.size());
        }
        return rs.iterator().next();
    }

    @Override
    public <E> List<E> selectForList(SQL sql, Object params, Class<E> requiredType) {
        String sqlStr = sql.toString();
        SqlParameterSource sqlParams = sqlParams(params);
        logger.debug("select sql: {} \n params: {}", sql, sqlParams);
        if (BeanUtils.isSimpleProperty(requiredType)) {
            return namedParameterJdbcOperations.queryForList(sqlStr, sqlParams, requiredType);
        }
        AnnotationRowMapper<E> rowMapper = new AnnotationRowMapper<>(requiredType);
        return namedParameterJdbcOperations.query(sqlStr, sqlParams, rowMapper);
    }

    @Override
    public <E> List<E> selectForPage(SQL sql, Object params, int pageNum, int pageSize, Class<E> requiredType) {
        String pageSQL = SQLBuilder.pageSQL(sql, getDbName(), pageNum, pageSize);
        SqlParameterSource sqlParams = sqlParams(params);
        logger.debug("page sql: {} \n params: {}", sql, sqlParams);
        if (BeanUtils.isSimpleProperty(requiredType)) {
            return namedParameterJdbcOperations.queryForList(pageSQL, sqlParams, requiredType);
        }
        AnnotationRowMapper<E> rowMapper = new AnnotationRowMapper<>(requiredType);
        return namedParameterJdbcOperations.query(pageSQL, sqlParams, rowMapper);
    }

    @Override
    public long count(SQL sql, Object params) {
        String countSql = sql.toCountSql();
        SqlParameterSource sqlParams = sqlParams(params);
        logger.debug("count sql: {} \n params: {}", countSql, sqlParams);
        Long count = namedParameterJdbcOperations.queryForObject(countSql, sqlParams, Long.class);
        return count == null ? 0L : count;
    }

    @Override
    public int update(SQL sql, Object params) {
        String updateSql = sql.toString();
        SqlParameterSource sqlParams = sqlParams(params);
        logger.debug("update sql: {} \n params: {}", updateSql, sqlParams);
        return namedParameterJdbcOperations.update(updateSql, sqlParams);
    }


    public String getDbName() {
        if (StringUtils.isEmpty(dbName)) {
            dbName = namedParameterJdbcOperations.getJdbcOperations()
                    .execute((ConnectionCallback<String>) con -> con.getMetaData().getDatabaseProductName());
        }
        return dbName;
    }

    public void setDbName(String dbName) {
        this.dbName = dbName;
    }


    public NamedParameterJdbcOperations getJdbcOperations() {
        return namedParameterJdbcOperations;
    }
}

File: yurt-data/yurt-data-jdbc/src/main/kotlin/com/github/liulus/yurt/jdbc/SQLExecutorImpl.kt
package com.github.liulus.yurt.jdbc

import org.springframework.beans.BeanUtils
import org.springframework.dao.IncorrectResultSizeDataAccessException
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.jdbc.core.namedparam.BeanPropertySqlParameterSource
import org.springframework.jdbc.core.namedparam.EmptySqlParameterSource
import org.springframework.jdbc.core.namedparam.MapSqlParameterSource
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcOperations
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate
import org.springframework.jdbc.core.namedparam.SqlParameterSource
import org.springframework.jdbc.support.GeneratedKeyHolder
import org.springframework.jdbc.support.KeyHolder
import org.springframework.util.Assert
import org.springframework.util.CollectionUtils
import org.springframework.util.ReflectionUtils
import org.springframework.util.StringUtils
import java.util.Objects
import javax.sql.DataSource

/**
 *
 * @author liulu
 * @version V1.0
 * @since 2020/11/5
 */
class SQLExecutorImpl(
        var namedParameterJdbcOperations: NamedParameterJdbcOperations
) : SQLExecutor {

    var dbName: String = ""
        get() {
            if (StringUtils.isEmpty(field)) {
                val jdbcOperations = namedParameterJdbcOperations.jdbcOperations
                if (jdbcOperations is JdbcTemplate) {
                    dbName = jdbcOperations.dataSource.connection.metaData.databaseProductName
                }
            }
            return field
        }

    constructor(dataSource: DataSource) : this(NamedParameterJdbcTemplate(dataSource))

    override fun <E> insert(entity: E): Long {
        requireNotNull(entity) { "insert with entity can not be null" }
        val entityClass = entity.javaClass
        val sql = SQLBuilder.insertSQL(entity)

        val tableMetadata = TableMetadata.forClass(entityClass)
        val keyHolder: KeyHolder = GeneratedKeyHolder()
        val sqlParams = sqlParams(entity)
        val insert: Int = namedParameterJdbcOperations.update(sql.toString(), sqlParams, keyHolder)
        val keyPs = BeanUtils.getPropertyDescriptor(entityClass, tableMetadata.idField)
        // set key property
        if (keyPs != null) {
            ReflectionUtils.invokeMethod(keyPs.writeMethod, entity,
                    Objects.requireNonNull(keyHolder.key).toLong())
        }
        return insert.toLong()
    }

    override fun <E> updateIgnoreNull(entity: E): Int {
        requireNotNull(entity) { "update with entity can not be null" }
        val sql = SQLBuilder.updateSQL(entity, true).toString()
        return namedParameterJdbcOperations.update(sql, sqlParams(entity))
    }

    override fun <E> deleteById(eClass: Class<E>, id: Long): Int {
        Assert.notNull(id, "id can not be null")
        val sql = SQLBuilder.deleteSQL(eClass).toString()
        return namedParameterJdbcOperations.update(sql, sqlParams(id))
    }

    override fun <E> selectById(eCLass: Class<E>, id: Long): E? {
        Assert.notNull(id, "id can not be null")
        val tableMetadata = TableMetadata.forClass(eCLass)
        val sql = SQL().SELECT(tableMetadata.columns).FROM(tableMetadata.tableName)
                .WHERE("${tableMetadata.idColumn} = :param")
        return selectForObject(sql, id, eCLass)
    }

    override fun <E> selectByIds(eClass: Class<E>, ids: Collection<Long>): List<E> {
        if (CollectionUtils.isEmpty(ids)) {
            return emptyList()
        }
        val tableMetadata = TableMetadata.forClass(eClass)
        val sql = SQL().SELECT(tableMetadata.columns).FROM(tableMetadata.tableName)
                .WHERE("${tableMetadata.idColumn} in :param")
        return selectForList(sql, ids, eClass)
    }


    override fun <E> selectForObject(sql: SQL, params: Any?, requiredType: Class<E>): E? {
        val rs = selectForList(sql, params, requiredType)
        if (CollectionUtils.isEmpty(rs)) {
            return null
        }
        if (rs.size > 1) {
            throw IncorrectResultSizeDataAccessException(1, rs.size)
        }
        return rs.iterator().next()
    }

    override fun <E> selectForList(sql: SQL, params: Any?, requiredType: Class<E>): List<E> {
        if (BeanUtils.isSimpleProperty(requiredType)) {
            return namedParameterJdbcOperations.queryForList(sql.toString(), sqlParams(params), requiredType)
        }
        val rowMapper = AnnotationRowMapper(requiredType)
        return namedParameterJdbcOperations.query(sql.toString(), sqlParams(params), rowMapper)

    }

    override fun count(sql: SQL, params: Any?): Long {
        return namedParameterJdbcOperations.queryForObject(sql.toCountSql(), sqlParams(params), Long::class.java) ?: 0L
    }

    override fun update(sql: SQL, params: Any?): Int {
        return namedParameterJdbcOperations.update(sql.toString(), sqlParams(params))
    }

    private fun <E> sqlParams(params: E): SqlParameterSource {
        if (params == null) {
            return EmptySqlParameterSource.INSTANCE
        }
        if (params is Collection<*>
                || BeanUtils.isSimpleProperty(params.javaClass)) {
            return MapSqlParameterSource("param", params)
        }
        if (params is Map<*, *>) {
            if (params.isEmpty()) {
                return EmptySqlParameterSource.INSTANCE
            }
            val parameterSource = MapSqlParameterSource()
            params.forEach { parameterSource.addValue(it.key as String, it.value) }
            return parameterSource
        }
        if (params is SqlParameterSource) {
            return params
        }
        return BeanPropertySqlParameterSource(params)
    }


}


